File: backend/src/tools/registry.ts
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";
import { env } from "../lib/env.js";
import { parseToolRegistryYaml } from "../lib/yaml.js";

export type ToolMeta = {
  command: string[];
  risk?: string;
  requires_target?: boolean;
  timeout?: number;
};

type ToolRegistryConfig = {
  tools?: Record<string, ToolMeta>;
};

export class ToolRegistry {
  private readonly tools: Record<string, ToolMeta>;

  constructor() {
    const filePath = path.join(env.configDir, "tool-registry.yaml");
    const raw = parseToolRegistryYaml(readFileSync(filePath, "utf8")) as ToolRegistryConfig;
    this.tools = raw.tools ?? {};
  }

  private commandAvailable(command: string[]) {
    const binary = command[0];
    if (!binary) return false;
    if (binary.includes("/")) return existsSync(binary);
    const paths = (process.env.PATH ?? "").split(path.delimiter).filter(Boolean);
    return paths.some((dir) => existsSync(path.join(dir, binary)));
  }

  list() {
    return Object.entries(this.tools).map(([name, meta]) => ({
      name,
      risk: meta.risk,
      requires_target: meta.requires_target ?? false,
      timeout: meta.timeout ?? 30,
      available: this.commandAvailable(meta.command),
      binary: meta.command[0] ?? ""
    }));
  }

  get(name: string) {
    return this.tools[name];
  }
}

File: backend/app/tools/registry.py
from pathlib import Path
import yaml
from app.core.config import CONFIG_DIR

class ToolRegistry:
    def __init__(self):
        self.raw = yaml.safe_load((Path(CONFIG_DIR) / "tool-registry.yaml").read_text())
        self.tools = self.raw.get("tools", {})

    def list(self):
        return [{"name": name, **{k:v for k,v in meta.items() if k != "command"}} for name, meta in self.tools.items()]

    def get(self, name):
        return self.tools.get(name)
